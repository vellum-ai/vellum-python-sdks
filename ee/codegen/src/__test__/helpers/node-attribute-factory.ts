import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
