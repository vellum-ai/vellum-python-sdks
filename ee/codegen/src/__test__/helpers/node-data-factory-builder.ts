import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
