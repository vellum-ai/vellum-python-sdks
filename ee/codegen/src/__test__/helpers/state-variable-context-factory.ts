import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
