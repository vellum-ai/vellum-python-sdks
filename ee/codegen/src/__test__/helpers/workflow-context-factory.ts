import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
