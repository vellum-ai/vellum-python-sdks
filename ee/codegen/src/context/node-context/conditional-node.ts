import { VellumVariableType } from "vellum-ai/api";

import { BaseNodeContext } from "src/context/node-context/base";
import { PortContext } from "src/context/port-context";
import { ConditionalNode } from "src/types/vellum";

export class ConditionalNodeContext extends BaseNodeContext<ConditionalNode> {
  baseNodeClassName = "ConditionalNode";
  baseNodeDisplayClassName = "BaseConditionalNodeDisplay";

  protected getNodeOutputNamesById(): Record<string, string> {
    return {};
  }

  protected getNodeOutputTypesById(): Record<string, VellumVariableType> {
    return {};
  }

  createPortContexts(): PortContext[] {
    return this.nodeData.data.conditions.map(
      (condition, idx) =>
        new PortContext({
          workflowContext: this.workflowContext,
          nodeContext: this,
          portId: condition.sourceHandleId,
          portName: `branch_${idx + 1}`,
          isDefault: false,
        })
    );
  }
}
