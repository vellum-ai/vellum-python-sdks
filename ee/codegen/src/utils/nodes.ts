import { VellumError } from "vellum-ai/errors";

import { WorkflowNode } from "src/types/vellum";

export function getNodeLabel(nodeData: WorkflowNode): string {
  switch (nodeData.type) {
    case "GENERIC":
      return nodeData.definition?.name ?? nodeData.label ?? "Generic Node";
    default:
      return nodeData.data.label;
  }
}

export function isUnaryOperator(operator: string): boolean {
  return operator === "null" || operator === "notNull";
}

export function isVellumErrorWithDetail(
  error: unknown
): error is VellumError & { body: { detail: string } } {
  return (
    error instanceof VellumError &&
    typeof error.body === "object" &&
    error.body !== null &&
    "detail" in error.body &&
    typeof error.body.detail === "string"
  );
}
