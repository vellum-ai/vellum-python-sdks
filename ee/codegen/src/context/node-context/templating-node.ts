import { VellumVariableType } from "vellum-ai/api";

import { BaseNodeContext } from "src/context/node-context/base";
import { PortContext } from "src/context/port-context";
import { TemplatingNode } from "src/types/vellum";

export class TemplatingNodeContext extends BaseNodeContext<TemplatingNode> {
  baseNodeClassName = "TemplatingNode";
  baseNodeDisplayClassName = "BaseTemplatingNodeDisplay";
  isCore = true;

  public getNodeOutputNamesById(): Record<string, string> {
    const errorOutputId = this.getErrorOutputId();
    return {
      [this.nodeData.data.outputId]: "result",
      ...(errorOutputId ? { [errorOutputId]: "error" } : {}),
    };
  }

  protected getNodeOutputTypesById(): Record<string, VellumVariableType> {
    return {
      [this.nodeData.data.outputId]: this.nodeData.data.outputType,
      ...(this.nodeData.data.errorOutputId
        ? { [this.nodeData.data.errorOutputId]: "ERROR" }
        : {}),
    };
  }

  createPortContexts(): PortContext[] {
    return [
      new PortContext({
        workflowContext: this.workflowContext,
        nodeContext: this,
        portId: this.nodeData.data.sourceHandleId,
      }),
    ];
  }
}
