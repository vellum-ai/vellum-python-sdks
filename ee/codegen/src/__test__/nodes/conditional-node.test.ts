import { Writer } from "@fern-api/python-ast/core/Writer";
import { v4 as uuid4 } from "uuid";
import { beforeEach } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { inputVariableContextFactory } from "src/__test__/helpers/input-variable-context-factory";
import {
  conditionalNodeFactory,
  conditionalNodeWithNullOperatorFactory,
  templatingNodeFactory,
} from "src/__test__/helpers/node-data-factories";
import { createNodeContext, WorkflowContext } from "src/context";
import { ConditionalNodeContext } from "src/context/node-context/conditional-node";
import { TemplatingNodeContext } from "src/context/node-context/templating-node";
import { ConditionalNode } from "src/generators/nodes/conditional-node";
import {
  ConditionalNode as ConditionalNodeType,
  WorkflowNodeType,
} from "src/types/vellum";

describe("ConditionalNode", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: ConditionalNode;

  beforeEach(async () => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    const nodeData = conditionalNodeFactory({ includeElif: true });

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "d2287fee-98fb-421c-9464-e54d8f70f046",
          key: "field",
          type: "STRING",
        },
        workflowContext,
      })
    );

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData,
    })) as ConditionalNodeContext;

    node = new ConditionalNode({
      workflowContext,
      nodeContext,
    });
  });

  it("getNodeFile", async () => {
    node.getNodeFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("getNodeDisplayFile", async () => {
    node.getNodeDisplayFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });
});

describe("ConditionalNode with invalid uuid for field and value node input ids", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: ConditionalNode;

  beforeEach(async () => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    const nodeData = constructConditionalNodeWithInvalidUUID();

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "d2287fee-98fb-421c-9464-e54d8f70f046",
          key: "field",
          type: "STRING",
        },
        workflowContext,
      })
    );

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData,
    })) as ConditionalNodeContext;

    node = new ConditionalNode({
      workflowContext,
      nodeContext,
    });
  });

  it("getNodeFile", async () => {
    node.getNodeFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("getNodeDisplayFile", async () => {
    node.getNodeDisplayFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  function constructConditionalNodeWithInvalidUUID(): ConditionalNodeType {
    return {
      id: "b81a4453-7b80-41ea-bd55-c62df8878fd3",
      type: WorkflowNodeType.CONDITIONAL,
      data: {
        label: "Conditional Node",
        targetHandleId: "842b9dda-7977-47ad-a322-eb15b4c7069d",
        conditions: [
          {
            id: "8d0d8b56-6c17-4684-9f16-45dd6ce23060",
            type: "IF",
            sourceHandleId: "63345ab5-1a4d-48a1-ad33-91bec41f92a5",
            data: {
              id: "fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
              rules: [
                {
                  id: "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc",
                  rules: [],
                  fieldNodeInputId:
                    "2cb6582e-c329-4952-8598-097830b766c7|cf63d0ad-5e52-4031-a29f-922e7004cdd8",
                  operator: "=",
                  valueNodeInputId:
                    "b51eb7cd-3e0a-4b42-a269-d58ebc3e0b04|51315413-f47c-4d7e-bc94-bd9e7862043d",
                },
              ],
              combinator: "AND",
            },
          },
          {
            id: "ea63ccd5-3fe3-4371-ba3c-6d3ec7ca2b60",
            type: "ELSE",
            sourceHandleId: "14a8b603-6039-4491-92d4-868a4dae4c15",
          },
        ],
        version: "2",
      },
      inputs: [
        {
          id: "2cb6582e-c329-4952-8598-097830b766c7|cf63d0ad-5e52-4031-a29f-922e7004cdd8",
          key: "2cb6582e-c329-4952-8598-097830b766c7|cf63d0ad-5e52-4031-a29f-922e7004cdd8",
          value: {
            rules: [
              {
                type: "INPUT_VARIABLE",
                data: {
                  inputVariableId: "d2287fee-98fb-421c-9464-e54d8f70f046",
                },
              },
            ],
            combinator: "OR",
          },
        },
        {
          id: "b51eb7cd-3e0a-4b42-a269-d58ebc3e0b04|51315413-f47c-4d7e-bc94-bd9e7862043d",
          key: "b51eb7cd-3e0a-4b42-a269-d58ebc3e0b04|51315413-f47c-4d7e-bc94-bd9e7862043d",
          value: {
            rules: [
              {
                type: "INPUT_VARIABLE",
                data: {
                  inputVariableId: "d2287fee-98fb-421c-9464-e54d8f70f046",
                },
              },
            ],
            combinator: "OR",
          },
        },
      ],
      displayData: {
        width: 480,
        height: 180,
        position: {
          x: 2247.2797390213086,
          y: 30.917121251477084,
        },
      },
    };
  }
});

describe("ConditionalNode with null operator", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: ConditionalNode;

  beforeEach(async () => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    const templatingNode = templatingNodeFactory();
    const nodeData = conditionalNodeWithNullOperatorFactory({
      nodeOutputReference: {
        nodeId: templatingNode.id,
        outputId: templatingNode.data.outputId,
      },
    });

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "d2287fee-98fb-421c-9464-e54d8f70f046",
          key: "field",
          type: "STRING",
        },
        workflowContext,
      })
    );

    (await createNodeContext({
      workflowContext,
      nodeData: templatingNode,
    })) as TemplatingNodeContext;

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData,
    })) as ConditionalNodeContext;

    node = new ConditionalNode({
      workflowContext,
      nodeContext,
    });
  });

  it("getNodeFile", async () => {
    node.getNodeFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("getNodeDisplayFile", async () => {
    node.getNodeDisplayFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });
});

describe("ConditionalNode with incorrect rule id references", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: ConditionalNode;

  beforeEach(async () => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    const invalidNodeData = constructInvalidConditionalNode();

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "d2287fee-98fb-421c-9464-e54d8f70f046",
          key: "field",
          type: "STRING",
        },
        workflowContext,
      })
    );

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData: invalidNodeData,
    })) as ConditionalNodeContext;

    node = new ConditionalNode({
      workflowContext,
      nodeContext,
    });
  });

  it("getNodeFile should throw error", async () => {
    try {
      node.getNodeFile().write(writer);
      await writer.toStringFormatted();
    } catch (error: unknown) {
      if (error instanceof Error) {
        expect(error.message).toBe(
          "Could not find input field key given ruleId: ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc on rule index: 0 on condition index: 0 for node: Conditional Node"
        );
      } else {
        throw new Error("Unexpected error type");
      }
    }
  });

  function constructInvalidConditionalNode(): ConditionalNodeType {
    return {
      id: "b81a4453-7b80-41ea-bd55-c62df8878fd3",
      type: WorkflowNodeType.CONDITIONAL,
      data: {
        label: "Conditional Node",
        targetHandleId: "842b9dda-7977-47ad-a322-eb15b4c7069d",
        conditions: [
          {
            id: "8d0d8b56-6c17-4684-9f16-45dd6ce23060",
            type: "IF",
            sourceHandleId: "63345ab5-1a4d-48a1-ad33-91bec41f92a5",
            data: {
              id: "fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
              rules: [
                {
                  id: "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc",
                  rules: [],
                  fieldNodeInputId: "2cb6582e-c329-4952-8598-097830b766c7",
                  operator: "=",
                  valueNodeInputId: "cf63d0ad-5e52-4031-a29f-922e7004cdd8",
                },
              ],
              combinator: "AND",
            },
          },
          {
            id: "ea63ccd5-3fe3-4371-ba3c-6d3ec7ca2b60",
            type: "ELSE",
            sourceHandleId: "14a8b603-6039-4491-92d4-868a4dae4c15",
          },
        ],
        version: "2",
      },
      inputs: [
        {
          id: "non-existent-id",
          key: "non-existent-rule-id.field",
          value: {
            rules: [
              {
                type: "INPUT_VARIABLE",
                data: {
                  inputVariableId: "d2287fee-98fb-421c-9464-e54d8f70f046",
                },
              },
            ],
            combinator: "OR",
          },
        },
      ],
      displayData: {
        width: 480,
        height: 180,
        position: {
          x: 2247.2797390213086,
          y: 30.917121251477084,
        },
      },
    };
  }
});

describe("Conditional Node with numeric operator casts rhs to NUMBER", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: ConditionalNode;

  beforeEach(async () => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "d2287fee-98fb-421c-9464-e54d8f70f046",
          key: "rhs",
          type: "NUMBER",
        },
        workflowContext,
      })
    );

    const nodeData: ConditionalNodeType = {
      id: "b81a4453-7b80-41ea-bd55-c62df8878fd3",
      type: WorkflowNodeType.CONDITIONAL,
      data: {
        label: "Conditional Node",
        targetHandleId: "842b9dda-7977-47ad-a322-eb15b4c7069d",
        conditions: [
          {
            id: "8d0d8b56-6c17-4684-9f16-45dd6ce23060",
            type: "IF",
            sourceHandleId: "63345ab5-1a4d-48a1-ad33-91bec41f92a5",
            data: {
              id: "fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
              rules: [
                {
                  id: "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc",
                  rules: [],
                  fieldNodeInputId: "2cb6582e-c329-4952-8598-097830b766c7",
                  operator: ">",
                  valueNodeInputId: "cf63d0ad-5e52-4031-a29f-922e7004cdd8",
                },
              ],
              combinator: "AND",
            },
          },
        ],
        version: "2",
      },
      inputs: [
        {
          id: "2cb6582e-c329-4952-8598-097830b766c7",
          key: "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc.field",
          value: {
            rules: [
              {
                type: "INPUT_VARIABLE",
                data: {
                  inputVariableId: "d2287fee-98fb-421c-9464-e54d8f70f046",
                },
              },
            ],
            combinator: "OR",
          },
        },
        {
          id: "cf63d0ad-5e52-4031-a29f-922e7004cdd8",
          key: "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc.value",
          value: {
            rules: [
              {
                type: "CONSTANT_VALUE",
                data: {
                  type: "STRING",
                  value: "0.5",
                },
              },
            ],
            combinator: "OR",
          },
        },
      ],
      displayData: {
        width: 480,
        height: 180,
        position: {
          x: 2247.2797390213086,
          y: 30.917121251477084,
        },
      },
    };

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData,
    })) as ConditionalNodeContext;

    node = new ConditionalNode({
      workflowContext,
      nodeContext,
    });
  });

  it("getNodeFile", async () => {
    node.getNodeFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });
});

describe("Conditional Node warning cases", () => {
  let writer: Writer;

  beforeEach(async () => {
    writer = new Writer();
  });

  it("getNodeFile should be resilient to lhs referencing a non-existent node", async () => {
    const workflowContext = workflowContextFactory({ strict: false });

    const referenceNodeId = uuid4();
    const nodeData = conditionalNodeFactory({
      // Non-existent node output reference
      inputReferenceId: uuid4(),
      inputReferenceNodeId: referenceNodeId,
    });

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData,
    })) as ConditionalNodeContext;

    const node = new ConditionalNode({
      workflowContext,
      nodeContext,
    });

    node.getNodeFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();

    // Ideally, we reduce the number of warnings to 1 in the future
    const errors = workflowContext.getErrors();
    expect(errors.length).toBe(2);
    expect(errors[0]?.message).toBe(
      `Failed to find node with id '${referenceNodeId}'`
    );
    expect(errors[1]?.message).toBe(
      `Node Conditional Node is missing required left-hand side input field for rule: 0 in condition: 0`
    );
    expect(errors[0]?.severity).toBe("WARNING");
    expect(errors[1]?.severity).toBe("WARNING");
  });
});
