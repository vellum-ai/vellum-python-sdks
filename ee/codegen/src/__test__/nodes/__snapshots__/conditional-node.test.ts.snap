// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Conditional Node with numeric operator casts rhs to NUMBER > getNodeFile 1`] = `
"from vellum.workflows.nodes.displayable import ConditionalNode as BaseConditionalNode
from vellum.workflows.ports import Port

from ..inputs import Inputs


class ConditionalNode(BaseConditionalNode):
    class Ports(BaseConditionalNode.Ports):
        branch_1 = Port.on_if(Inputs.rhs.greater_than(0.5))
"
`;

exports[`ConditionalNode > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseConditionalNodeDisplay
from vellum_ee.workflows.display.nodes.vellum.conditional_node import (
    ConditionId,
    RuleIdMap,
)
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ...nodes.conditional_node import ConditionalNode


class ConditionalNodeDisplay(BaseConditionalNodeDisplay[ConditionalNode]):
    label = "Conditional Node"
    node_id = UUID("b81a4453-7b80-41ea-bd55-c62df8878fd3")
    target_handle_id = UUID("842b9dda-7977-47ad-a322-eb15b4c7069d")
    source_handle_ids = {
        0: UUID("63345ab5-1a4d-48a1-ad33-91bec41f92a5"),
        1: UUID("2c03f27f-ea64-42fc-8a6c-383550c58ae4"),
        2: UUID("14a8b603-6039-4491-92d4-868a4dae4c15"),
    }
    rule_ids = [
        RuleIdMap(
            id="fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
            lhs=RuleIdMap(
                id="ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc",
                lhs=None,
                rhs=None,
                field_node_input_id=None,
                value_node_input_id=None,
            ),
            rhs=None,
            field_node_input_id=None,
            value_node_input_id=None,
        ),
        RuleIdMap(
            id="fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
            lhs=RuleIdMap(
                id="ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc",
                lhs=None,
                rhs=None,
                field_node_input_id=None,
                value_node_input_id=None,
            ),
            rhs=None,
            field_node_input_id=None,
            value_node_input_id=None,
        ),
    ]
    condition_ids = [
        ConditionId(
            id="8d0d8b56-6c17-4684-9f16-45dd6ce23060",
            rule_group_id="fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
        ),
        ConditionId(
            id="e63c3933-ef86-451f-88bc-d7ea7dce4310",
            rule_group_id="fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
        ),
        ConditionId(id="ea63ccd5-3fe3-4371-ba3c-6d3ec7ca2b60", rule_group_id=None),
    ]
    node_input_ids_by_name = {
        "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc.field": UUID(
            "2cb6582e-c329-4952-8598-097830b766c7"
        ),
        "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc.value": UUID(
            "cf63d0ad-5e52-4031-a29f-922e7004cdd8"
        ),
    }
    port_displays = {}
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=2247.2797390213086, y=30.917121251477084),
        width=480,
        height=180,
    )
"
`;

exports[`ConditionalNode > getNodeFile 1`] = `
"from vellum.workflows.nodes.displayable import ConditionalNode as BaseConditionalNode
from vellum.workflows.ports import Port

from ..inputs import Inputs


class ConditionalNode(BaseConditionalNode):
    class Ports(BaseConditionalNode.Ports):
        branch_1 = Port.on_if(Inputs.field.equals("testtest"))
        branch_2 = Port.on_elif(Inputs.field.is_not_null())
        branch_3 = Port.on_else()
"
`;

exports[`ConditionalNode with invalid uuid for field and value node input ids > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseConditionalNodeDisplay
from vellum_ee.workflows.display.nodes.vellum.conditional_node import (
    ConditionId,
    RuleIdMap,
)
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ...nodes.conditional_node import ConditionalNode


class ConditionalNodeDisplay(BaseConditionalNodeDisplay[ConditionalNode]):
    label = "Conditional Node"
    node_id = UUID("b81a4453-7b80-41ea-bd55-c62df8878fd3")
    target_handle_id = UUID("842b9dda-7977-47ad-a322-eb15b4c7069d")
    source_handle_ids = {
        0: UUID("63345ab5-1a4d-48a1-ad33-91bec41f92a5"),
        1: UUID("14a8b603-6039-4491-92d4-868a4dae4c15"),
    }
    rule_ids = [
        RuleIdMap(
            id="fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
            lhs=RuleIdMap(
                id="ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc",
                lhs=None,
                rhs=None,
                field_node_input_id=None,
                value_node_input_id=None,
            ),
            rhs=None,
            field_node_input_id=None,
            value_node_input_id=None,
        )
    ]
    condition_ids = [
        ConditionId(
            id="8d0d8b56-6c17-4684-9f16-45dd6ce23060",
            rule_group_id="fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
        ),
        ConditionId(id="ea63ccd5-3fe3-4371-ba3c-6d3ec7ca2b60", rule_group_id=None),
    ]
    node_input_ids_by_name = {
        "2cb6582e-c329-4952-8598-097830b766c7|cf63d0ad-5e52-4031-a29f-922e7004cdd8": "2cb6582e-c329-4952-8598-097830b766c7|cf63d0ad-5e52-4031-a29f-922e7004cdd8",
        "b51eb7cd-3e0a-4b42-a269-d58ebc3e0b04|51315413-f47c-4d7e-bc94-bd9e7862043d": "b51eb7cd-3e0a-4b42-a269-d58ebc3e0b04|51315413-f47c-4d7e-bc94-bd9e7862043d",
    }
    port_displays = {}
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=2247.2797390213086, y=30.917121251477084),
        width=480,
        height=180,
    )
"
`;

exports[`ConditionalNode with invalid uuid for field and value node input ids > getNodeFile 1`] = `
"from vellum.workflows.nodes.displayable import ConditionalNode as BaseConditionalNode
from vellum.workflows.ports import Port

from ..inputs import Inputs


class ConditionalNode(BaseConditionalNode):
    class Ports(BaseConditionalNode.Ports):
        branch_1 = Port.on_if(Inputs.field.equals(Inputs.field))
        branch_2 = Port.on_else()
"
`;

exports[`ConditionalNode with null operator > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseConditionalNodeDisplay
from vellum_ee.workflows.display.nodes.vellum.conditional_node import (
    ConditionId,
    RuleIdMap,
)
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ...nodes.conditional_node import ConditionalNode


class ConditionalNodeDisplay(BaseConditionalNodeDisplay[ConditionalNode]):
    label = "Conditional Node"
    node_id = UUID("b81a4453-7b80-41ea-bd55-c62df8878fd3")
    target_handle_id = UUID("842b9dda-7977-47ad-a322-eb15b4c7069d")
    source_handle_ids = {0: UUID("63345ab5-1a4d-48a1-ad33-91bec41f92a5")}
    rule_ids = [
        RuleIdMap(
            id="fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
            lhs=RuleIdMap(
                id="ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc",
                lhs=None,
                rhs=None,
                field_node_input_id=None,
                value_node_input_id=None,
            ),
            rhs=RuleIdMap(
                id="ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cd",
                lhs=None,
                rhs=None,
                field_node_input_id=None,
                value_node_input_id=None,
            ),
            field_node_input_id=None,
            value_node_input_id=None,
        )
    ]
    condition_ids = [
        ConditionId(
            id="8d0d8b56-6c17-4684-9f16-45dd6ce23060",
            rule_group_id="fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
        )
    ]
    node_input_ids_by_name = {
        "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc.field": UUID(
            "2cb6582e-c329-4952-8598-097830b766c7"
        ),
        "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cd.field": UUID(
            "2cb6582e-c329-4952-8598-097830b766c8"
        ),
    }
    port_displays = {}
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=2247.2797390213086, y=30.917121251477084),
        width=480,
        height=180,
    )
"
`;

exports[`ConditionalNode with null operator > getNodeFile 1`] = `
"from vellum.workflows.nodes.displayable import ConditionalNode as BaseConditionalNode
from vellum.workflows.ports import Port

from ..inputs import Inputs
from .templating_node import TemplatingNode


class ConditionalNode(BaseConditionalNode):
    class Ports(BaseConditionalNode.Ports):
        branch_1 = Port.on_if(
            Inputs.field.is_null() and TemplatingNode.Outputs.result.is_nil()
        )
"
`;
